armeabi  is not work on armv7 devices

For smaller size, We add 
```
ndk {
    abiFilters "armeabi"
}
```
in My app's build.gradle file, It means always compile armeabi folder, so use armeabi so if on armv7 devices, That's work for all tripartite library.

But when use mapbox, map is empty on armv7 devices if only use armeabi so like above.

I want to know why so lib of armeabi is not work on armv7 devices?

Thank you very much. @bleege

Just confirming that the TestApp shows the following crash on a Nexus 5 if we include the filter indicated above:

```java
03-14 17:23:14.458 12684-12684/? I/art: Late-enabling -Xcheck:jni
03-14 17:23:14.513 12684-12684/com.mapbox.mapboxsdk.testapp W/System: ClassLoader referenced unknown path: /data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm
03-14 17:23:14.533 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err: java.io.FileNotFoundException: /jacoco.exec: open failed: EROFS (Read-only file system)
03-14 17:23:14.533 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err:     at libcore.io.IoBridge.open(IoBridge.java:452)
03-14 17:23:14.533 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err:     at java.io.FileOutputStream.<init>(FileOutputStream.java:87)
03-14 17:23:14.533 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err:     at org.jacoco.agent.rt.internal_773e439.output.FileOutput.openFile(FileOutput.java:67)
03-14 17:23:14.533 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err:     at org.jacoco.agent.rt.internal_773e439.output.FileOutput.startup(FileOutput.java:49)
03-14 17:23:14.533 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err:     at org.jacoco.agent.rt.internal_773e439.Agent.startup(Agent.java:122)
03-14 17:23:14.533 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err:     at org.jacoco.agent.rt.internal_773e439.Agent.getInstance(Agent.java:50)
03-14 17:23:14.533 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err:     at org.jacoco.agent.rt.internal_773e439.Offline.<clinit>(Offline.java:31)
03-14 17:23:14.533 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err:     at com.mapbox.mapboxsdk.testapp.MapboxApplication.$jacocoInit(MapboxApplication.java)
03-14 17:23:14.533 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err:     at com.mapbox.mapboxsdk.testapp.MapboxApplication.<init>(MapboxApplication.java)
03-14 17:23:14.533 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err:     at java.lang.Class.newInstance(Native Method)
03-14 17:23:14.533 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err:     at android.app.Instrumentation.newApplication(Instrumentation.java:996)
03-14 17:23:14.533 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err:     at android.app.Instrumentation.newApplication(Instrumentation.java:981)
03-14 17:23:14.533 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err:     at android.app.LoadedApk.makeApplication(LoadedApk.java:573)
03-14 17:23:14.534 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err:     at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4680)
03-14 17:23:14.534 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err:     at android.app.ActivityThread.-wrap1(ActivityThread.java)
03-14 17:23:14.534 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err:     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1405)
03-14 17:23:14.534 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:102)
03-14 17:23:14.534 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err:     at android.os.Looper.loop(Looper.java:148)
03-14 17:23:14.534 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err:     at android.app.ActivityThread.main(ActivityThread.java:5417)
03-14 17:23:14.534 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err:     at java.lang.reflect.Method.invoke(Native Method)
03-14 17:23:14.534 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err:     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
03-14 17:23:14.534 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
03-14 17:23:14.534 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err: Caused by: android.system.ErrnoException: open failed: EROFS (Read-only file system)
03-14 17:23:14.534 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err:     at libcore.io.Posix.open(Native Method)
03-14 17:23:14.534 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err:     at libcore.io.BlockGuardOs.open(BlockGuardOs.java:186)
03-14 17:23:14.534 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err:     at libcore.io.IoBridge.open(IoBridge.java:438)
03-14 17:23:14.534 12684-12684/com.mapbox.mapboxsdk.testapp W/System.err: 	... 21 more
03-14 17:23:14.548 12684-12704/com.mapbox.mapboxsdk.testapp D/LeakCanary: Could not attempt cleanup, leak storage not writable.
03-14 17:23:14.624 12684-12684/com.mapbox.mapboxsdk.testapp W/LocationServices: Location Permissions Not Granted Yet.  Try again after requesting.
03-14 17:23:14.635 12684-12684/com.mapbox.mapboxsdk.testapp D/AndroidRuntime: Shutting down VM
03-14 17:23:14.636 12684-12684/com.mapbox.mapboxsdk.testapp E/AndroidRuntime: FATAL EXCEPTION: main
                                                                              Process: com.mapbox.mapboxsdk.testapp, PID: 12684
                                                                              java.lang.RuntimeException: Unable to start activity ComponentInfo{com.mapbox.mapboxsdk.testapp/com.mapbox.mapboxsdk.testapp.MainActivity}: android.view.InflateException: Binary XML file line #32: Binary XML file line #32: Error inflating class com.mapbox.mapboxsdk.maps.MapView
                                                                                  at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2416)
                                                                                  at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476)
                                                                                  at android.app.ActivityThread.-wrap11(ActivityThread.java)
                                                                                  at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344)
                                                                                  at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                                  at android.os.Looper.loop(Looper.java:148)
                                                                                  at android.app.ActivityThread.main(ActivityThread.java:5417)
                                                                                  at java.lang.reflect.Method.invoke(Native Method)
                                                                                  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                                                  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
                                                                               Caused by: android.view.InflateException: Binary XML file line #32: Binary XML file line #32: Error inflating class com.mapbox.mapboxsdk.maps.MapView
                                                                                  at android.view.LayoutInflater.inflate(LayoutInflater.java:539)
                                                                                  at android.view.LayoutInflater.inflate(LayoutInflater.java:423)
                                                                                  at android.view.LayoutInflater.inflate(LayoutInflater.java:374)
                                                                                  at android.support.v7.app.AppCompatDelegateImplV7.setContentView(AppCompatDelegateImplV7.java:256)
                                                                                  at android.support.v7.app.AppCompatActivity.setContentView(AppCompatActivity.java:109)
                                                                                  at com.mapbox.mapboxsdk.testapp.MainActivity.onCreate(MainActivity.java:103)
                                                                                  at android.app.Activity.performCreate(Activity.java:6251)
                                                                                  at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1107)
                                                                                  at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2369)
                                                                                  at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476) 
                                                                                  at android.app.ActivityThread.-wrap11(ActivityThread.java) 
                                                                                  at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344) 
                                                                                  at android.os.Handler.dispatchMessage(Handler.java:102) 
                                                                                  at android.os.Looper.loop(Looper.java:148) 
                                                                                  at android.app.ActivityThread.main(ActivityThread.java:5417) 
                                                                                  at java.lang.reflect.Method.invoke(Native Method) 
                                                                                  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
                                                                                  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) 
                                                                               Caused by: android.view.InflateException: Binary XML file line #32: Error inflating class com.mapbox.mapboxsdk.maps.MapView
                                                                                  at android.view.LayoutInflater.createView(LayoutInflater.java:645)
                                                                                  at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:764)
                                                                                  at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:704)
                                                                                  at android.view.LayoutInflater.rInflate(LayoutInflater.java:835)
                                                                                  at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:798)
                                                                                  at android.view.LayoutInflater.rInflate(LayoutInflater.java:838)
                                                                                  at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:798)
                                                                                  at android.view.LayoutInflater.rInflate(LayoutInflater.java:838)
                                                                                  at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:798)
                                                                                  at android.view.LayoutInflater.rInflate(LayoutInflater.java:838)
                                                                                  at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:798)
                                                                                  at android.view.LayoutInflater.inflate(LayoutInflater.java:515)
                                                                                  at android.view.LayoutInflater.inflate(LayoutInflater.java:423) 
                                                                                  at android.view.LayoutInflater.inflate(LayoutInflater.java:374) 
                                                                                  at android.support.v7.app.AppCompatDelegateImplV7.setContentView(AppCompatDelegateImplV7.java:256) 
                                                                                  at android.support.v7.app.AppCompatActivity.setContentView(AppCompatActivity.java:109) 
                                                                                  at com.mapbox.mapboxsdk.testapp.MainActivity.onCreate(MainActivity.java:103) 
                                                                                  at android.app.Activity.performCreate(Activity.java:6251) 
                                                                                  at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1107) 
                                                                                  at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2369) 
                                                                                  at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476) 
                                                                                  at android.app.ActivityThread.-wrap11(ActivityThread.java) 
                                                                                  at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344) 
                                                                                  at android.os.Handler.dispatchMessage(Handler.java:102) 
                                                                                  at android.os.Looper.loop(Looper.java:148) 
                                                                                  at android.app.ActivityThread.main(ActivityThread.java:5417) 
                                                                                  at java.lang.reflect.Method.invoke(Native Method) 
                                                                                  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
                                                                                  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) 
                                                                               Caused by: java.lang.reflect.InvocationTargetException
                                                                                  at java.lang.reflect.Constructor.newInstance(Native Method)
                                                                                  at android.view.LayoutInflater.createView(LayoutInflater.java:619)
                                                                                  at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:764) 
                                                                                  at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:704) 
                                                                                  at android.view.LayoutInflater.rInflate(LayoutInflater.java:835) 
                                                                                  at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:798) 
                                                                                  at android.view.LayoutInflater.rInflate(LayoutInflater.java:838) 
                                                                                  at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:798) 
                                                                                  at android.view.LayoutInflater.rInflate(LayoutInflater.java:838) 
                                                                                  at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:798) 
                                                                                  at android.view.LayoutInflater.rInflate(LayoutInflater.java:838) 
                                                                                  at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:798) 
                                                                                  at android.view.LayoutInflater.inflate(LayoutInflater.java:515) 
                                                                                  at android.view.LayoutInflater.inflate(LayoutInflater.java:423) 
                                                                                  at android.view.LayoutInflater.inflate(LayoutInflater.java:374) 
                                                                                  at android.support.v7.app.AppCompatDelegateImplV7.setContentView(AppCompatDelegateImplV7.java:256) 
                                                                                  at android.support.v7.app.AppCompatActivity.setContentView(AppCompatActivity.java:109) 
                                                                                  at com.mapbox.mapboxsdk.testapp.MainActivity.onCreate(MainActivity.java:103) 
                                                                                  at android.app.Activity.performCreate(Activity.java:6251) 
                                                                                  at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1107) 
                                                                                  at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2369) 
                                                                                  at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476) 
                                                                                  at android.app.ActivityThread.-wrap11(ActivityThread.java) 
                                                                                  at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344) 
                                                                                  at android.os.Handler.dispatchMessage(Handler.java:102) 
                                                                                  at android.os.Looper.loop(Looper.java:148) 
                                                                                  at android.app.ActivityThread.main(ActivityThread.java:5417) 
                                                                                  at java.lang.reflect.Method.invoke(Native Method) 
                                                                                  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
                                                                                  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) 
                                                                               Caused by: java.lang.UnsatisfiedLinkError: dalvik.system.PathClassLoader[DexPathList[[zip file "/data/app/com.mapbox.mapboxsdk.testapp-2/base.apk"],nativeLibraryDirectories=[/data/app/com.mapbox.mapboxsdk.testapp-2/lib/arm, /vendor/lib, /system/lib]]] couldn't find "libmapbox-gl.so"
                                                                                  at java.lang.Runtime.loadLibrary(Runtime.java:367)
                                                                                  at java.lang.System.loadLibrary(System.java:1076)
                                                                                  at com.mapbox.mapboxsdk.maps.NativeMapView.<clinit>(NativeMapView.java:44)
                                                                                  at com.mapbox.mapboxsdk.maps.MapView.initialize(MapView.java:193)
                                                                                  at com.mapbox.mapboxsdk.maps.MapView.<init>(MapView.java:157)
                                                                                  at java.lang.reflect.Constructor.newInstance(Native Method) 
                                                                                  at android.view.LayoutInflater.createView(LayoutInflater.java:619) 
                                                                                  at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:764) 
                                                                                  at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:704) 
                                                                                  at android.view.LayoutInflater.rInflate(LayoutInflater.java:835) 
                                                                                  at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:798) 
                                                                                  at android.view.LayoutInflater.rInflate(LayoutInflater.java:838) 
                                                                                  at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:798) 
                                                                                  at android.view.LayoutInflater.rInflate(LayoutInflater.java:838) 
                                                                                  at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:798) 
                                                                                  at android.view.LayoutInflater.rInflate(LayoutInflater.java:838) 
                                                                                  at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:798) 
                                                                                  at android.view.LayoutInflater.inflate(LayoutInflater.java:515) 
                                                                                  at android.view.LayoutInflater.inflate(LayoutInflater.java:423) 
                                                                                  at android.view.LayoutInflater.inflate(LayoutInflater.java:374) 
                                                                                  at android.support.v7.app.AppCompatDelegateImplV7.setContentView(AppCompatDelegateImplV7.java:256) 
                                                                                  at android.support.v7.app.AppCompatActivity.setContentView(AppCompatActivity.java:109) 
                                                                                  at com.mapbox.mapboxsdk.testapp.MainActivity.onCreate(MainActivity.java:103) 
                                                                                  at android.app.Activity.performCreate(Activity.java:6251) 
                                                                                  at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1107) 
                                                                                  at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2369) 
                                                                                  at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476) 
                                                                                  at android.app.ActivityThread.-wrap11(ActivityThread.java) 
                                                                                  at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344) 
                                                                                  at android.os.Handler.dispatchMessage(Handler.java:102) 
                                                                                  at android.os.Looper.loop(Looper.java:148) 
                                                                                  at android.app.ActivityThread.main(ActivityThread.java:5417) 
                                                                                  at java.lang.reflect.Method.invoke(Native Method) 
                                                                                  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
                                                                                  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) 

```

--- a/include/mbgl/storage/network_status.hpp
+++ b/include/mbgl/storage/network_status.hpp
@@ -1,6 +1,7 @@
 #pragma once
 
-#include <atomic>
+#include <mbgl/util/atomic.hpp>
+
 #include <mutex>
 #include <set>
 
@@ -26,7 +27,7 @@ public:
     static void Unsubscribe(util::AsyncTask* async);
 
 private:
-    static std::atomic<bool> online;
+    static util::Atomic<bool> online;
     static std::mutex mtx;
     static std::set<util::AsyncTask*> observers;
 };

--- a/include/mbgl/util/run_loop.hpp
+++ b/include/mbgl/util/run_loop.hpp
@@ -1,5 +1,6 @@
 #pragma once
 
+#include <mbgl/util/atomic.hpp>
 #include <mbgl/util/noncopyable.hpp>
 #include <mbgl/util/util.hpp>
 #include <mbgl/util/work_task.hpp>
@@ -9,7 +10,6 @@
 #include <utility>
 #include <queue>
 #include <mutex>
-#include <atomic>
 
 namespace mbgl {
 namespace util {
@@ -59,7 +59,7 @@ public:
     template <class Fn, class... Args>
     std::unique_ptr<AsyncRequest>
     invokeCancellable(Fn&& fn, Args&&... args) {
-        auto flag = std::make_shared<std::atomic<bool>>();
+        auto flag = std::make_shared<util::Atomic<bool>>();
         *flag = false;
 
         auto tuple = std::make_tuple(std::move(args)...);
@@ -77,7 +77,7 @@ public:
     template <class Fn, class Cb, class... Args>
     std::unique_ptr<AsyncRequest>
     invokeWithCallback(Fn&& fn, Cb&& callback, Args&&... args) {
-        auto flag = std::make_shared<std::atomic<bool>>();
+        auto flag = std::make_shared<util::Atomic<bool>>();
         *flag = false;
 
         // Create a lambda L1 that invokes another lambda L2 on the current RunLoop R, that calls
@@ -114,7 +114,7 @@ private:
     template <class F, class P>
     class Invoker : public WorkTask {
     public:
-        Invoker(F&& f, P&& p, std::shared_ptr<std::atomic<bool>> canceled_ = nullptr)
+        Invoker(F&& f, P&& p, std::shared_ptr<util::Atomic<bool>> canceled_ = nullptr)
           : canceled(std::move(canceled_)),
             func(std::move(f)),
             params(std::move(p)) {
@@ -148,7 +148,7 @@ private:
         }
 
         std::recursive_mutex mutex;
-        std::shared_ptr<std::atomic<bool>> canceled;
+        std::shared_ptr<util::Atomic<bool>> canceled;
 
         F func;
         P params;

--- a/platform/android/src/async_task.cpp
+++ b/platform/android/src/async_task.cpp
@@ -1,9 +1,9 @@
 #include "run_loop_impl.hpp"
 
 #include <mbgl/util/async_task.hpp>
+#include <mbgl/util/atomic.hpp>
 #include <mbgl/util/run_loop.hpp>
 
-#include <atomic>
 #include <functional>
 
 namespace mbgl {
@@ -45,7 +45,7 @@ private:
 
     // TODO: Use std::atomic_flag if we ever drop
     // support for ARMv5
-    std::atomic<bool> queued;
+    util::Atomic<bool> queued;
     std::function<void()> task;
 };
 

--- a/platform/android/src/async_task.cpp
+++ b/platform/android/src/async_task.cpp
@@ -12,7 +12,7 @@ namespace util {
 class AsyncTask::Impl : public RunLoop::Impl::Runnable {
 public:
     Impl(std::function<void()>&& fn)
-        : task(std::move(fn)) {
+        : queued(true), task(std::move(fn)) {
         loop->initRunnable(this);
     }
 
@@ -21,7 +21,8 @@ public:
     }
 
     void maySend() {
-        if (!queued.test_and_set()) {
+        if (queued) {
+            queued = false;
             loop->addRunnable(this);
         }
     }
@@ -32,7 +33,7 @@ public:
 
     void runTask() override {
         loop->removeRunnable(this);
-        queued.clear();
+        queued = true;
         task();
     }
 
@@ -42,7 +43,9 @@ private:
 
     RunLoop::Impl* loop = reinterpret_cast<RunLoop::Impl*>(RunLoop::getLoopHandle());
 
-    std::atomic_flag queued = ATOMIC_FLAG_INIT;
+    // TODO: Use std::atomic_flag if we ever drop
+    // support for ARMv5
+    std::atomic<bool> queued;
     std::function<void()> task;
 };
 

--- a/platform/android/src/run_loop_impl.hpp
+++ b/platform/android/src/run_loop_impl.hpp
@@ -2,10 +2,10 @@
 
 #include "jni.hpp"
 
+#include <mbgl/util/atomic.hpp>
 #include <mbgl/util/chrono.hpp>
 #include <mbgl/util/run_loop.hpp>
 
-#include <atomic>
 #include <list>
 #include <memory>
 #include <mutex>
@@ -47,7 +47,7 @@ private:
     bool detach = false;
 
     ALooper* loop = nullptr;
-    std::atomic<bool> running;
+    util::Atomic<bool> running;
 
     std::recursive_mutex mtx;
     std::list<Runnable*> runnables;

--- a/src/mbgl/geometry/glyph_atlas.hpp
+++ b/src/mbgl/geometry/glyph_atlas.hpp
@@ -2,6 +2,7 @@
 
 #include <mbgl/geometry/binpack.hpp>
 #include <mbgl/text/glyph_store.hpp>
+#include <mbgl/util/atomic.hpp>
 #include <mbgl/util/noncopyable.hpp>
 #include <mbgl/util/optional.hpp>
 #include <mbgl/gl/gl.hpp>
@@ -11,7 +12,6 @@
 #include <set>
 #include <unordered_map>
 #include <mutex>
-#include <atomic>
 
 namespace mbgl {
 
@@ -53,7 +53,7 @@ private:
     BinPack<uint16_t> bin;
     std::unordered_map<FontStack, std::map<uint32_t, GlyphValue>, FontStackHash> index;
     const std::unique_ptr<uint8_t[]> data;
-    std::atomic<bool> dirty;
+    util::Atomic<bool> dirty;
     mbgl::optional<gl::UniqueTexture> texture;
 };
 

--- a/src/mbgl/map/map.cpp
+++ b/src/mbgl/map/map.cpp
@@ -510,7 +510,7 @@ CameraOptions Map::cameraForLatLngs(const std::vector<LatLng>& latLngs, optional
         scaleY -= (padding->top + padding->bottom) / height;
     }
     double minScale = ::fmin(scaleX, scaleY);
-    double zoom = ::log2(getScale() * minScale);
+    double zoom = util::log2(getScale() * minScale);
     zoom = util::clamp(zoom, getMinZoom(), getMaxZoom());
 
     // Calculate the center point of a virtual bounds that is extended in all directions by padding.

--- a/src/mbgl/map/transform_state.cpp
+++ b/src/mbgl/map/transform_state.cpp
@@ -222,7 +222,7 @@ double TransformState::zoomScale(double zoom) const {
 }
 
 double TransformState::scaleZoom(double s) const {
-    return ::log2(s);
+    return util::log2(s);
 }
 
 double TransformState::worldSize() const {

--- a/src/mbgl/renderer/bucket.hpp
+++ b/src/mbgl/renderer/bucket.hpp
@@ -2,10 +2,9 @@
 
 #include <mbgl/gl/gl.hpp>
 #include <mbgl/renderer/render_pass.hpp>
+#include <mbgl/util/atomic.hpp>
 #include <mbgl/util/noncopyable.hpp>
 #include <mbgl/util/mat4.hpp>
-
-#include <atomic>
 
 #define BUFFER_OFFSET_0  ((GLbyte*)nullptr)
 #define BUFFER_OFFSET(i) ((BUFFER_OFFSET_0) + (i))
@@ -50,7 +49,7 @@ public:
     virtual void swapRenderData() {}
 
 protected:
-    std::atomic<bool> uploaded;
+    util::Atomic<bool> uploaded;
 
 };
 

--- a/src/mbgl/sprite/sprite_atlas.hpp
+++ b/src/mbgl/sprite/sprite_atlas.hpp
@@ -3,6 +3,7 @@
 #include <mbgl/geometry/binpack.hpp>
 #include <mbgl/gl/gl.hpp>
 #include <mbgl/gl/object_store.hpp>
+#include <mbgl/util/atomic.hpp>
 #include <mbgl/util/noncopyable.hpp>
 #include <mbgl/util/ptr.hpp>
 #include <mbgl/util/optional.hpp>
@@ -10,7 +11,6 @@
 #include <string>
 #include <map>
 #include <mutex>
-#include <atomic>
 #include <set>
 #include <array>
 
@@ -92,7 +92,7 @@ private:
     std::map<Key, Holder> images;
     std::set<std::string> uninitialized;
     std::unique_ptr<uint32_t[]> data;
-    std::atomic<bool> dirty;
+    util::Atomic<bool> dirty;
     bool fullUploadRequired = true;
     mbgl::optional<gl::UniqueTexture> texture;
     uint32_t filter = 0;

--- a/src/mbgl/storage/network_status.cpp
+++ b/src/mbgl/storage/network_status.cpp
@@ -9,7 +9,7 @@
 
 namespace mbgl {
 
-std::atomic<bool> NetworkStatus::online(true);
+util::Atomic<bool> NetworkStatus::online(true);
 std::mutex NetworkStatus::mtx;
 std::set<util::AsyncTask *> NetworkStatus::observers;
 

--- a/src/mbgl/style/bucket_parameters.hpp
+++ b/src/mbgl/style/bucket_parameters.hpp
@@ -3,9 +3,9 @@
 #include <mbgl/map/mode.hpp>
 #include <mbgl/tile/tile_id.hpp>
 #include <mbgl/style/filter.hpp>
+#include <mbgl/util/atomic.hpp>
 
 #include <functional>
-#include <atomic>
 
 namespace mbgl {
 
@@ -24,7 +24,7 @@ class BucketParameters {
 public:
     BucketParameters(const OverscaledTileID& tileID_,
                           const GeometryTileLayer& layer_,
-                          const std::atomic<bool>& obsolete_,
+                          const util::Atomic<bool>& obsolete_,
                           uintptr_t tileUID_,
                           bool& partialParse_,
                           SpriteStore& spriteStore_,
@@ -51,7 +51,7 @@ public:
 
     const OverscaledTileID& tileID;
     const GeometryTileLayer& layer;
-    const std::atomic<bool>& obsolete;
+    const util::Atomic<bool>& obsolete;
     uintptr_t tileUID;
     bool& partialParse;
     SpriteStore& spriteStore;

--- a/src/mbgl/text/glyph_pbf.hpp
+++ b/src/mbgl/text/glyph_pbf.hpp
@@ -2,10 +2,10 @@
 
 #include <mbgl/text/glyph.hpp>
 #include <mbgl/text/glyph_store.hpp>
+#include <mbgl/util/atomic.hpp>
 #include <mbgl/util/font_stack.hpp>
 #include <mbgl/util/noncopyable.hpp>
 
-#include <atomic>
 #include <functional>
 #include <string>
 #include <memory>
@@ -29,7 +29,7 @@ public:
     }
 
 private:
-    std::atomic<bool> parsed;
+    util::Atomic<bool> parsed;
     std::unique_ptr<AsyncRequest> req;
     GlyphStoreObserver* observer = nullptr;
 };

--- a/src/mbgl/tile/tile_worker.cpp
+++ b/src/mbgl/tile/tile_worker.cpp
@@ -24,7 +24,7 @@ TileWorker::TileWorker(const OverscaledTileID& id_,
                        SpriteStore& spriteStore_,
                        GlyphAtlas& glyphAtlas_,
                        GlyphStore& glyphStore_,
-                       const std::atomic<bool>& obsolete_,
+                       const util::Atomic<bool>& obsolete_,
                        const MapMode mode_)
     : id(id_),
       sourceID(std::move(sourceID_)),

--- a/src/mbgl/tile/tile_worker.hpp
+++ b/src/mbgl/tile/tile_worker.hpp
@@ -2,6 +2,7 @@
 
 #include <mbgl/map/mode.hpp>
 #include <mbgl/tile/tile_id.hpp>
+#include <mbgl/util/atomic.hpp>
 #include <mbgl/util/noncopyable.hpp>
 #include <mbgl/util/variant.hpp>
 #include <mbgl/util/ptr.hpp>
@@ -12,7 +13,6 @@
 #include <memory>
 #include <mutex>
 #include <list>
-#include <atomic>
 #include <unordered_map>
 
 namespace mbgl {
@@ -50,7 +50,7 @@ public:
                SpriteStore&,
                GlyphAtlas&,
                GlyphStore&,
-               const std::atomic<bool>&,
+               const util::Atomic<bool>&,
                const MapMode);
     ~TileWorker();
 
@@ -75,7 +75,7 @@ private:
     SpriteStore& spriteStore;
     GlyphAtlas& glyphAtlas;
     GlyphStore& glyphStore;
-    const std::atomic<bool>& obsolete;
+    const util::Atomic<bool>& obsolete;
     const MapMode mode;
 
     bool partialParse = false;

--- a/src/mbgl/tile/vector_tile_data.hpp
+++ b/src/mbgl/tile/vector_tile_data.hpp
@@ -3,9 +3,9 @@
 #include <mbgl/tile/tile_data.hpp>
 #include <mbgl/tile/tile_worker.hpp>
 #include <mbgl/text/placement_config.hpp>
+#include <mbgl/util/atomic.hpp>
 #include <mbgl/util/feature.hpp>
 
-#include <atomic>
 #include <memory>
 #include <unordered_map>
 
@@ -69,7 +69,7 @@ private:
     PlacementConfig targetConfig;
 
     // Used to signal the worker that it should abandon parsing this tile as soon as possible.
-    std::atomic<bool> obsolete { false };
+    util::Atomic<bool> obsolete { false };
 };
 
 } // namespace mbgl

--- a/src/mbgl/util/atomic.hpp
+++ b/src/mbgl/util/atomic.hpp
@@ -1,0 +1,45 @@
+#pragma once
+
+#include <atomic>
+#include <mutex>
+
+namespace mbgl {
+namespace util {
+
+// std::atomic<bool> is implemented lock free which
+// is not supported by ARMv5 but the code generated
+// seems to be using that and not working at all,
+// thus, this naive implementation using mutexes.
+#if defined(__ANDROID__) && defined(__ARM_ARCH_5TE__)
+
+template <typename T>
+class Atomic {
+public:
+    Atomic() = default;
+    explicit Atomic(const T& data_) : data(data_) {}
+
+    void operator=(const T& other) {
+        std::lock_guard<std::mutex> lock(mtx);
+        data = other;
+    }
+
+    operator bool() const {
+        std::lock_guard<std::mutex> lock(mtx);
+
+        return data;
+    }
+
+private:
+    T data;
+    mutable std::mutex mtx;
+};
+
+#else
+
+template <typename T>
+using Atomic = std::atomic<T>;
+
+#endif
+
+} // namespace util
+} // namespace mbgl

--- a/src/mbgl/util/math.cpp
+++ b/src/mbgl/util/math.cpp
@@ -21,5 +21,15 @@ uint32_t ceil_log2(uint64_t x) {
     return y;
 }
 
+double log2(double x) {
+// log2() is producing wrong results on ARMv5 binaries
+// running on ARMv7+ CPUs.
+#if defined(__ANDROID__) && defined(__ARM_ARCH_5TE__)
+    return std::log(x) / 0.6931471805599453; // log(x) / log(2)
+#else
+    return ::log2(x);
+#endif
+}
+
 } // namespace util
 } // namespace mbgl

--- a/src/mbgl/util/math.hpp
+++ b/src/mbgl/util/math.hpp
@@ -110,5 +110,7 @@ T smoothstep(T edge0, T edge1, T x) {
 // (== number of bits required to store x)
 uint32_t ceil_log2(uint64_t x);
 
+double log2(double x);
+
 } // namespace util
 } // namespace mbgl

--- a/src/mbgl/util/thread.hpp
+++ b/src/mbgl/util/thread.hpp
@@ -6,6 +6,7 @@
 #include <utility>
 #include <functional>
 
+#include <mbgl/util/atomic.hpp>
 #include <mbgl/util/run_loop.hpp>
 #include <mbgl/util/thread_context.hpp>
 #include <mbgl/platform/platform.hpp>

